src/lib/SaferNames/Name.hs

```diff
@@ -4,288 +4,206 @@
+
+appendScope :: Scope n -> ScopeFrag n l -> Scope l
+appendScope (Scope m1) (ScopeFrag m2) =
+  Scope $ extendNameMap (injectNames (nameMapNames m2) m1) m2
 
 -- === common scoping patterns ===
 
 data Abs (binder::B) (body::E) (n::S) where
   Abs :: binder n l -> body l -> Abs binder body n
 
-data FreshAbs (binder::B) (body::E) (n::S) where
-  FreshAbs :: FreshExt n l -> binder n l -> body l -> FreshAbs binder body n
-
 data NestPair (b1::B) (b2::B) (n::S) (l::S) where
   NestPair :: b1 n l -> b2 l l' -> NestPair b1 b2 n l'
 
 data Nest (binder::B) (n::S) (l::S) where
   Nest  :: binder n h -> Nest binder h l -> Nest binder n l
   Empty ::                                  Nest binder n n
 
-type PlainBinderList = Nest PlainBinder
-
-infixl 7 :>
-data AnnBinderP (b::B) (ann ::E) (n::S) (l::S) =
-  (:>) (b n l) (ann n)
-  deriving (Show)
+type EmptyAbs b = Abs b UnitE :: E
+pattern EmptyAbs :: b n l -> EmptyAbs b n
+pattern EmptyAbs bs = Abs bs UnitE
+
+-- === Injections and projections ===
+
+newtype Ext (n::S) (l::S) = ExtVal { fromExtVal :: NameSet (n:=>:l) }
+
+instance Category Ext where
+  id = ExtVal emptyNameSetFrag
+  ext1 . ext2 = ExtVal $ concatNameSets (fromExtVal ext2) (fromExtVal ext1)
+
+class BindsNames (b :: B) where
+  toExtVal :: b n l -> Ext n l
+
+instance BindsNames Ext where
+  toExtVal = id
+
+toNameSet :: BindsNames b => b n l -> NameSet (n:=>:l)
+toNameSet b = fromExtVal $ toExtVal b
+
+inject :: BindsNames b => InjectableE e => Distinct l => b n l -> e n -> e l
+inject ext x = injectNames (toNameSet ext) x
+
+-- like inject, but uses the SopeReader monad for its `Distinct` proof
+injectM :: ScopeReader m => BindsNames b => InjectableE e => b n l -> e n -> m l (e l)
+injectM b e = withDistinct $ return $ inject b e
+
+fromConstAbs :: ScopeReader m => MonadErr1 m => BindsNames b => HasNamesE e
+             => Abs b e n -> m n (e n)
+fromConstAbs (Abs b e) = do
+  let t = newNameTraversal \name ->
+             case projectName (toNameSet b) name of
+               Left name' -> Just name'
+               Right _    -> Nothing
+  scope <- askScope
+  case traverseNamesE scope t e of
+    Nothing -> throw EscapedNameErr ""
+    Just e' -> return e'
+
+-- TODO: seems silly to have to have the static annotation here. I think we need
+-- to have a finer-grained scope reader class hierarchy so we're not forced to
+-- supply static information when it's not needed.
+toConstAbs :: InjectableE s => Typeable s => ScopeReader m => HasNamesE e
+           => s n -> e n -> m n (Abs (NameBinder s) e n)
+toConstAbs ann body =
+  withFreshM ann \b -> do
+    body' <- injectM b body
+    return $ Abs b body'
+
+-- === type classes for traversing names ===
+
+data NameTraversal m i o where
+  NameTraversal :: (forall s. Name s i -> m (Name s o)) -- monadic function for free vars
+                -> EnvFrag Name i i' o                  -- renaming local vars
+                -> NameTraversal m i' o
```

Comment:
The fact that you use `i` in the `data` declaration, and the `i'` for the same thing in the constructor definition (and `i` to mean something else) is _extremely_ confusing. Can you please rename those?

Reply:
Also can you please add docs for what `i` and `o` are supposed to be? Are they the scopes of free variables? "renaming local vars" is also insufficient for me to understand what's going on. Is `i'` the local scope? Why is it wrapped as an existential thing in a traversal?

Reply:
If so, perhaps rename `i'` to `l` for local? Or just use longer names here? No one forces us to use one-letter aliases and while it's justifiable in many function signatures, they do have special meaning here

Reply:
I use this naming convention a lot, so maybe it's worth writing a note about it. It looks something like this:
```
                             | before/above binders | after/below binders
input  (i.e. pre-traversal)  |         i            |        i'
output (i.e. post-traversal) |         o            |        o'
```
Does that picture help?

Reply:
The local vars / free vars thing is a distraction. It happens to be the case that we composition of function-and-env splits along a free/local vars boundary -- we start with a function that we want to apply to free vars, and then we rename vars as we traverse the expression and go under binders -- but that's not actually important.